File: apps/web/src/lib/research-display-contract.ts
import { object, type Row } from './research-reader-model';

export const RESEARCH_FIELD_LABELS = {
  title: '제목', korean_summary: '요약', key_points: '핵심 주장', catalysts: '촉매',
  risks: '반대 근거·위험', invalidation_conditions: '판단을 바꿀 조건', source_document_ids: '리서치 입력 문서',
} as const;
export type ResearchField = keyof typeof RESEARCH_FIELD_LABELS;
export type ResearchDisplayIssue = { invalid: ResearchField[]; unavailable: ResearchField[]; metadataInvalid: boolean };
const fields = Object.keys(RESEARCH_FIELD_LABELS) as ResearchField[];
function fieldList(value: unknown): value is ResearchField[] {
  return Array.isArray(value) && value.every(item => typeof item === 'string' && fields.includes(item as ResearchField))
    && new Set(value).size === value.length;
}
/** Quality is a typed producer contract, not an accuracy/confidence score. */
export function researchDisplayIssue(research: Row): ResearchDisplayIssue | null {
  if (!Object.hasOwn(research, 'data_quality')) return null; // Legacy payload: no validation claim.
  const quality = object(research.data_quality);
  const invalid = quality.invalid_fields, unavailable = quality.unavailable_fields;
  if (quality.policy !== 'equity_display_contract_v1' || !fieldList(invalid) || !fieldList(unavailable)
    || invalid.some(field => unavailable.includes(field))
    || quality.status !== (invalid.length ? 'invalid_fields' : unavailable.length ? 'partial' : 'complete')) {
    return { invalid: [...fields], unavailable: [], metadataInvalid: true };
  }
  return invalid.length || unavailable.length ? { invalid: [...invalid], unavailable: [...unavailable], metadataInvalid: false } : null;
}
/** Null means unknown to the existing readers. Only an actual valid [] means empty. */
export function projectResearchForReader(value: unknown): Row {
  const raw = object(value), issue = researchDisplayIssue(raw);
  if (!issue) return raw;
  const projected = { ...raw };
  for (const field of [...issue.invalid, ...issue.unavailable]) projected[field] = null;
  return projected;
}

File: src/stockanalysis/frontend/research_display_contract.py
"""Validate stored narrative fields before a display adapter stringifies them.

This is not financial scoring or semantic verification. Keep the established API
string/list shapes, with explicit field-quality metadata for rejected/unavailable
input. Never salvage only the valid elements of a malformed claim/source list.
"""
from __future__ import annotations

from typing import Any

POLICY = 'equity_display_contract_v1'
TEXT_FIELDS = ('title', 'korean_summary')
CLAIM_FIELDS = ('key_points', 'catalysts', 'risks', 'invalidation_conditions')
MAX_DATABASE_ID = 9223372036854775807


def _document_id(value: object) -> bool:
    if type(value) is int:
        return 0 < value <= MAX_DATABASE_ID
    if type(value) is not str:
        return False
    # Stored artifact source inventories contain database IDs. The existing API
    # builder converts them to source-document-N, without going through JS float.
    return (0 < len(value) <= 19 and value.isascii() and value.isdecimal()
            and value[0] != '0' and int(value) <= MAX_DATABASE_ID)


def prepare_equity_display(artifact: dict[str, Any]) -> tuple[dict[str, Any], dict[str, object]]:
    safe = dict(artifact)
    invalid: list[str] = []
    unavailable: list[str] = []
    for field in TEXT_FIELDS:
        value = artifact.get(field)
        if value is None:
            unavailable.append(field)
            safe[field] = ''
        elif type(value) is not str:
            invalid.append(field)
            safe[field] = ''
        else:
            safe[field] = value
    for field in (*CLAIM_FIELDS, 'source_document_ids'):
        value = artifact.get(field)
        if value is None:
            unavailable.append(field)
            safe[field] = []
        elif type(value) is not list or not all(
            _document_id(item) if field == 'source_document_ids' else type(item) is str
            for item in value
        ):
            invalid.append(field)
            safe[field] = []
        else:
            safe[field] = list(value)
    quality: dict[str, object] = {
        'policy': POLICY,
        'status': 'invalid_fields' if invalid else 'partial' if unavailable else 'complete',
        'invalid_fields': invalid,
        'unavailable_fields': unavailable,
    }
    return safe, quality
